cvat/apps/engine/ddln/utils.py

```diff
@@ -111,23 +78,73 @@ def get_sequence_id_mapping(task_name):
     return {f.parent.name: f.read_text().strip() for f in task_dir.glob("*/ddln_id")}
 
 
-YML_TEMPLATE = """sources:
-  - {ddln_id}
-date: {curr_date}
-team:
-  - group: {group}
-  - mapping: {map_file}
-phabricator: {task_name}
-tool:
-  - name: CVAT
-    version: 2.3
-  - name: merge tool https://git-ng.daedalean.ai/daedalean/exp-devtools/src/branch/master/annotations/multi/process_annotations_msq.py
-    version: {merger_version}
-invalid:
-{invalid_data}comment:
-quality:
-recommendations: """
-
-INVALID_SEQUENCE_TEMPLATE = """  - dataset_id:  {dataset_id}
-{frame_rows}    reason: no agreement
-"""
+class OrderedDumper(yaml.Dumper):
+    pass
+
+
+def _represent_ordered_dict(dumper, data):
+    return dumper.represent_mapping(yaml.resolver.BaseResolver.DEFAULT_MAPPING_TAG, data.items())
+
+
+def _represent_none(dumper, data):
+    return dumper.represent_scalar('tag:yaml.org,2002:null', '')
+
+
+OrderedDumper.add_representer(OrderedDict, _represent_ordered_dict)
+OrderedDumper.add_representer(type(None), _represent_none)
+
+
+class DdlnYamlWriter:
+    def __init__(self, task_name, add_merger_info=True):
+        self.task_name = guess_task_name(task_name)
+        self.annotation_request_id = get_annotation_request_id(self.task_name)
+        self.id_by_seq_name = get_sequence_id_mapping(self.task_name)
+        self._add_merger_info = add_merger_info
+
+    def get_warnings(self, sequence_names=()):
+        warnings = []
+        if not self.annotation_request_id:
+            warnings.append("Failure while obtaining annotation request id")
+        if not all(name in self.id_by_seq_name for name in sequence_names):
+            warnings.append("Failure while getting sequence to dataset-id mapping")
+        return warnings
+
+    def write_metadata(self, file, date=None):
+        if date is None:
+            date = datetime.date.today()
+        annotation_request_id = self.annotation_request_id or ''
+        group = settings.ANNOTATION_TEAM
+        map_file = "task_mapping.csv"
+
+        data = OrderedDict([
+            ("sources", [annotation_request_id]),
+            ("date", date),
+            ("team", OrderedDict([("group", group), ("mapping", map_file)])),
+            ("phabricator", self.task_name),
+            ("tool", [
+                OrderedDict([("name", "CVAT"), ("version", 2.3)]),
```

Comment:
would be nice at some point to record git hash of the current CVAT version, not just `2.3`

Reply:
I cannot save CVAT git hash in settings (like I do for `exp-devtools`) because modifying and committing that file creates new commit so the stored hash becomes invalid. I cannot get the value at runtime because `.git` directory is not copied into docker image so cvat is no longer a git repository.

Reply:
doesn't have to happen in this PR, just for the future:
this is solved by adding another command when creating the docker image, that explicitly writes the git hash into some file or variable